hack/externalnode/install-vm.sh

```diff
@@ -120,10 +130,37 @@ Restart=on-failure
 WantedBy=multi-user.target
 EOF
     systemctl daemon-reload
-    systemctl enable antrea-agent
-    echo "Starting antrea-agent service"
-    systemctl start antrea-agent
-    systemctl status antrea-agent
+    systemctl enable "$ANTREA_AGENT"
+    echo "Starting ${ANTREA_AGENT} service"
+    systemctl start "$ANTREA_AGENT"
+    systemctl status "$ANTREA_AGENT"
+}
+
+start_antrea_cleanup_service() {
+    echo "Copying $BASH_SOURCE to ${AGENT_CONF_PATH}/install-vm.sh"
+    cp "$BASH_SOURCE" "${AGENT_CONF_PATH}/install-vm.sh"
+    chmod +x "${AGENT_CONF_PATH}/install-vm.sh"
+    cat >/etc/systemd/system/antrea-cleanup.service << EOF
+[Unit]
+Description="antrea cleanup as a systemd service"
+Requires=${OVS_VSWITCHD}
+After=network.target ${OVS_VSWITCHD}
+[Service]
+ExecStart="${AGENT_CONF_PATH}/install-vm.sh" --clear-ovs --ovs-bridge $OVS_BRIDGE
```

Comment:
does it mean the cleanup will always be executed when ovs-vswitchd starts or only be executed on system start or user manual start?
asking because I want to understand why this is not a step of the antrea agent service but a separate service.

Reply:
This mean that antrea-cleanup service requires ovs-vswitchd service to be up and running before it can run.
This is required, since the cleanup service is responsible to deleting OVS bridge.
The service will be executed only during system start up.

Antrea-agent service can run independently, since it will run ovs-* related commands only when it is connected to OVS.

Reply:
The antrea-cleanup service stops the antrea-agent service. Once it is stopped then it deletes the bridge. So there would not be race condition while deleting the bridge.

After reboot, the interfaces names are renamed back. So any configuration in ovs is invalid, as shown in the issue.
The cleanup is a must after every reboot of VM.

Reply:
On Windows, upon reboot VM loses network connectivity. antrea-agent will not be able to connect to Antrea/K8s API server to receive ExternalNode configuration. 
Antrea-agent proceeds with agent initialization only when it detects a matching ExternalNode corresponding to the environment variable NODE_NAME or hostname.

On Linux, adding additional logic while adding ExternalNode could be done. 
@wenyingd  can comment more about this.

Reply:
This looks more complex than handling error port cleanup in antrea-agent: there would be one service for antrea-agent, another service for antrea's cleanup, and when the latter is executed, it will stop the former to ensure correct order.
On Windows, when the OVS ports are in error state, I think the physical adapter is not attached to OVS bridge so it should have no network connectivity issue, no? You mean we must delete the error ports then physical adapter can work?

Reply:
On Windows, the virtual adapter (Ethernet) created by antrea-agent, will never get an IP address. This is because the OVS does not recognise the virtual adapter(Ethernet) device. 

To get it to work, we have to follow the steps listed in this doc to recover.
https://github.com/antrea-io/antrea/blob/main/docs/external-node.md#known-issues

`
PS C:\Windows\system32> Get-netadapter

Name                      InterfaceDescription                    ifIndex Status
----                      --------------------                    ------- ----- 
Ethernet 2                Mellanox ConnectX-3 Virtual Function...       8 Up    
Ethernet                  Hyper-V Virtual Ethernet Adapter              7 Up    
Ethernet~                 Microsoft Hyper-V Network Adapter             3 Up 
`

PS C:\Windows\system32> ovs-vsctl show                                          
a1d5caf9-1a3c-4d77-9500-edda35587586                                            
    Bridge br-int
        datapath_type: system                                                   
        Port Ethernet                                                           
            Interface Ethernet
                type: internal                                                  
                error: "could not add network device Ethernet to ofproto (Invalid argument)"                                                                    
        Port "Ethernet~"                                                        
            Interface "Ethernet~"                                               
    ovs_version: "2.16.3"

Reply:
I still do not get why the cleanup cannot be in the antrea-agent service.

If the cleanup brings the network down, could we check if it is really needed. It is not good to break the connections when restart a service.

Reply:
> I still do not get why the cleanup cannot be in the antrea-agent service.
> 
> If the cleanup brings the network down, could we check if it is really needed. It is not good to break the connections when restart a service.

Without this patch, the current behaviour on windows is that:  Upon system reboot, there is no network connectivity. One has to login to the VM by attaching a secondary interface or by other means(serial console), and then clean up ovs bridge, remove VmSwitch. Only then the network connectivity is restored.

Reply:
I don't prefer to place the cleanup logic inside antrea-agent code (golang): although the cleanup requirement exist both in Linux and Windows, the scenarios are different: 1) Linux connectivity is not impacted, since the VM interface is not attached on OVS bridge because of name conflict after startup (Linux removes the internal interface created by antrea, and rename the interface back), so OVS can not find the interfaces defined in OVSDB. 2) Windows host connectivity is lost after reboot (as Anand has printed), and antrea-agent is blocking at connecting to kubernetes APIServer to get the ExternalNode resource, so antrea-agent does not have a chance to modify OVS at this time. 

If the concern is the conflict between two services, shall we move a precheck and cleanup inside antrea-agent service ( not antrea-agent golang code). Then the workflow is, when we start antrea-agent service on host, a pre-check is run to check if any errors exist on OVS interfaces, if yes, the service clean up OVS configurations, and then run antrea-agent bits.  Since the error happen only after a startup, the cleanup operation is run at that time. And for a long-running host, even user manually restart antrea-agent service, no cleanup is taken as there is no error existing on OVS, so it would not block anything.

Reply:
@wenyingd  discussed offline with @tnqn, and we decided to take the above the approach suggested by Wenying.
Updated the PR. Please help with review.

Reply:
The new design sounds good to me.